src/coinjoin/client.cpp

```diff
@@ -1885,7 +1885,14 @@ void DoCoinJoinMaintenance(CConnman& connman, CBlockPolicyEstimator& fee_estimat
         coinJoinClientQueueManager->DoMaintenance();
     }
 
-    for (const auto& pair : coinJoinClientManagers) {
+    for (auto pair : coinJoinClientManagers->raw()) {
         pair.second->DoMaintenance(connman, fee_estimator, mempool);
     }
 }
+
+void CJClientManager::Add(CWallet& wallet) {
+    m_imap.emplace(
+        std::make_pair(wallet.GetName(),
+                       std::make_shared<CCoinJoinClientManager>(wallet, m_mn_sync))
```

Comment:
I am not sure about this part. Can you help me to figure out:

so far as CCoinJoinClientManager is added through shared pointer, can you show that it will be destroyed BEFORE `wallet`?  Otherwise if somehow that manager is not destroyed during some internal calls or something like DoMaintainence - it can be catastrophic.

Reply:
While they're all now `unique_ptr` now, from my recollection, CoinJoin entities expects wallet logic to be initialized and ready _before_ wallets are registered.

Keeping that in mind, the wallets need to be unregistered (they are registered as `chain_clients` as you can see [here](https://github.com/dashpay/dash/blob/9f7322b34a317b55608630a88d88ea3bcff114d7/src/wallet/init.cpp#L181-L191)), then CoinJoin destroyed, then wallet instances destroyed and that's been done [here](https://github.com/dashpay/dash/blob/180482b4f323727c9628561412cd964e2fde561b/src/init.cpp#L382-L386)